src/node_modules/app/machines/explore/search.options.js

```diff
@@ -0,0 +1,289 @@
+	return ctx;
+};
+
+const loadFieldStatsFromCache = ctx => updateFieldStats(
+	ctx,
+	{data: get(ctx.statsCache)[get(ctx.statsCacheKey)]}
+);
+
+const isInStatsCache = ctx =>
+	get(ctx.statsCacheKey) in get(ctx.statsCache);
+
+const setMenuActive = ctx => {
+	ctx.isFieldsMenuActive.set(true);
+	return ctx;
+}
+
+const setMenuInactive = ctx => {
+	ctx.isFieldsMenuActive.set(false);
+	return ctx;
+}
+
+const updateSearchQuery = (ctx, {detail}) => {
+	ctx.searchQuery.set(detail.trim().toLowerCase());
+	return ctx;
+}
+
+function selectNextSearchField (ctx) {
+	const selectedFieldName = get(ctx.selectedFieldName);
+	const nextFieldNames = get(ctx.nextFieldNames);
+	const nextFieldId = nextFieldNames[selectedFieldName];
+	if (nextFieldId) {
+		ctx.userSelection.set(nextFieldId);
+		ctx.selectedFieldName.set(computeFieldSelection(
+			get(ctx.fieldStats),
+			get(ctx.userSelection)
+		));
+	}
+	return ctx;
+}
+
+function selectPrevSearchField (ctx) {
+	const selectedFieldName = get(ctx.selectedFieldName);
+	const prevFieldNames = get(ctx.prevFieldNames);
+	const prevFieldId = prevFieldNames[selectedFieldName];
+	if (prevFieldId) {
+		ctx.userSelection.set(prevFieldId);
+		ctx.selectedFieldName.set(computeFieldSelection(
+			get(ctx.fieldStats),
+			get(ctx.userSelection)
+		));
+	}
+	return ctx;
+}
+
+function updateCurrentField (ctx, {detail}) {
+	ctx.userSelection.set(detail);
+	ctx.selectedFieldName.set(computeFieldSelection(
+		get(ctx.fieldStats),
+		get(ctx.userSelection)
+	));
+	return ctx;
+}
+
+// Suggestions utils
+const computeSuggestionsQuery = (requestedFieldName, term) => ({
+	size: 0,
+	query: {
+		term: {
+			[requestedFieldName]: term
+		}
+	},
+	aggs: {
+		[requestedFieldName]: {
+			significant_text: {
+				field: requestedFieldName,
+				// TODO include/exclude according to ES version
+				include: stopWords,
+				min_doc_count: 1,
+				size: 20
+			}
+		}
+	}
+});
+
+// Suggestions options
+const isSuggestionsRequestIncomplete = ctx =>
+	!(Boolean(get(ctx.searchQuery)) && Boolean(get(ctx.selectedFieldName)));
+
+const suggestionsRequest = ctx => request(
+	'POST',
+	getSearchURL(get(selectedDataset)),
+	{
+		data: computeSuggestionsQuery(
+			get(ctx.selectedFieldName),
+			get(ctx.searchQuery)
+		)
+	}
+);
+
+const updateSuggestionsCache = (ctx, {data}) => {
+	ctx.suggestionsCache.update(_.setKey(get(ctx.suggestionsCacheKey), data));
+	return ctx;
+};
+
+const updateSuggestions = (ctx, {data}) => {
+	const aggs = data.aggregations[get(ctx.selectedFieldName)];
+	const buckets =
+		(aggs && aggs.buckets || [])
+		.filter(sugg => sugg.key !== get(ctx.searchQuery));
+	const sortedBuckets = _.sort(buckets, [
+		_.sorterDesc(_.getKey('doc_count')),
+		_.getKey('key')]
+	);
+	const suggestions =
+		sortedBuckets
+		.map(sugg => `${sugg.key} (${sugg.doc_count})`);
+	if (suggestions.length === 0) {
+		suggestions.push('-- no suggestions found --')
+	}
+	ctx.suggestions.set(suggestions);
+	return ctx;
+}
```

Comment:
I'd try something a bit more functional like: 

```suggestion
const makeGetSuggestionsBy = searchQuery => _.pipe([
	_.filterWith(sugg => sugg.key !== searchQuery),
	_.sortWith([
		_.sorterDesc(_.getKey('doc_count')),
		_.getKey('key')
	]),
	_.mapWith(sugg => `${sugg.key} (${sugg.doc_count})`),
	_.when(isIterableEmpty, _.always(['-- no suggestions found --']))
]);

const updateSuggestions = (ctx, {data}) => {
	const selectedFieldName = get(ctx.selectedFieldName);
	const searchQuery = get(ctx.searchQuery);

	const aggs = data.aggregations[selectedFieldName];
	const buckets = aggs && aggs.buckets || [];
	const makeSuggestions = makeGetSuggestionsBy(searchQuery);
	const suggestions = makeSuggestions(buckets);
	ctx.suggestions.set(suggestions);
	
	return ctx;
}
```

[1] In general we should try to get all stores values first, so that it's clear what we're going to use, isolating computation so that we can read what's going on at a glance possibly.